packages/@eventual/aws-cdk/src/socket-service.ts

```diff
@@ -0,0 +1,179 @@
+          Object.entries(this.sockets as Record<string, Socket>).map(
+            ([name, socket]) =>
+              [
+                name,
+                {
+                  http: socket.gatewayStage.url.replace("wss://", "https://"),
+                  wss: socket.gatewayStage.url,
+                } satisfies SocketUrls,
+              ] as const
+          )
+        )
+      ),
+  } as const;
+
+  private addEnvs(func: Function, ...envs: (keyof typeof this.ENV_MAPPINGS)[]) {
+    envs.forEach((env) => func.addEnvironment(env, this.ENV_MAPPINGS[env]()));
+  }
+
+  public configureInvokeSocketEndpoints(func: Function) {
+    this.grantInvokeSocketEndpoints(func);
+    this.addEnvs(func, ENV_NAMES.SOCKET_URLS);
+  }
+
+  public grantInvokeSocketEndpoints(grantable: IGrantable) {
+    // generally we want to compute the grants, but apigateway urls use the generated appid and not the name
+    Object.values<Socket>(this.sockets).map((s) =>
+      s.gateway.grantManageConnections(grantable)
+    );
+  }
+}
+
+interface SocketProps {
+  serviceProps: SocketsProps<any>;
+  socketService: SocketService<any>;
+  socket: SocketFunction;
+  local: ServiceLocal | undefined;
+}
+
+export interface ISocket {
+  grantPrincipal: IPrincipal;
+  gateway: IWebSocketApi;
+  gatewayStage: WebSocketStage;
+  handler: Function;
+}
+
+class Socket extends Construct implements EventualResource, ISocket {
+  public grantPrincipal: IPrincipal;
+  public gateway: WebSocketApi;
+  public gatewayStage: WebSocketStage;
+  public handler: Function;
+
+  constructor(scope: Construct, props: SocketProps) {
+    const socketName = props.socket.spec.name;
+
+    super(scope, socketName);
+
+    this.handler = new ServiceFunction(this, "DefaultHandler", {
+      build: props.serviceProps.build,
+      bundledFunction: props.socket,
+      functionNameSuffix: `socket-${socketName}-default`,
+      serviceName: props.serviceProps.serviceName,
+      defaults: {
+        timeout: Duration.minutes(1),
+        environment: {
+          [ENV_NAMES.SOCKET_NAME]: socketName,
+          ...props.serviceProps.environment,
+        },
+      },
+      runtimeProps: props.socket.spec,
+      overrides: props.serviceProps.overrides?.[socketName],
+    });
+
+    configureWorkerCalls(props.serviceProps, this.handler);
+
+    this.gateway = new WebSocketApi(this, "Gateway", {
+      apiName: socketServiceSocketName(
+        props.serviceProps.serviceName,
+        socketName
+      ),
+      defaultRouteOptions: {
+        // https://stackoverflow.com/a/72716478
+        // must create one integration per...
+        integration: new WebSocketLambdaIntegration("default", this.handler),
+      },
+      connectRouteOptions: {
+        integration: new WebSocketLambdaIntegration("Connect", this.handler),
+        authorizer: new WebSocketNoneAuthorizer(),
+      },
+      disconnectRouteOptions: {
+        integration: new WebSocketLambdaIntegration("Disconnect", this.handler),
```

Comment:
Dang, those LambdaIntegrations. I wonder if we can work around that?

Not high priority, but gah-damn that is bad